Fix thread-safety issues with enumerating ResourceManager.

madelson opened the pull request:
Concurrently enumerating a ResourceManager while also calling GetString()
and similar methods was prone to both transient errors and deadlock.

The transient errors were caused by RuntimeResourceSet calling internal
methods on ResourceReader that did not properly lock. Now, all exposed
methods on the Reader are thread-safe.

The deadlock was called by inconsistent lock ordering between
ResourceReader.ResourceEnumerator and RuntimeResourceSet which both lock
on the RuntimeResourceSet's cache and on the ResourceReader itself. Now,
the enumerator does not need to take both locks at the same time.

Fix #74052
Fix #74868

madelson committed:
Fix thread-safety issues with enumerating ResourceManager.

Concurrently enumerating a ResourceManager while also calling GetString()
and similar methods was prone to both transient errors and deadlock.

The transient errors were caused by RuntimeResourceSet calling internal
methods on ResourceReader that did not properly lock. Now, all exposed
methods on the Reader are thread-safe.

The deadlock was called by inconsistent lock ordering between
ResourceReader.ResourceEnumerator and RuntimeResourceSet which both lock
on the RuntimeResourceSet's cache and on the ResourceReader itself. Now,
the enumerator does not need to take both locks at the same time.

Fix #74052
Fix #74868

madelson commented on src/libraries/System.Private.CoreLib/src/System/Resources/ResourceReader.Core.cs:
```diff
@@ -55,6 +55,8 @@ internal ResourceReader(Stream stream, Dictionary<string, ResourceLocator> resCa
             "the user to only get one error.")]
         private object DeserializeObject(int typeIndex)
         {
+            Debug.Assert(Monitor.IsEntered(this));
```
I added these to private methods that access `_store` per [this suggestion](https://github.com/dotnet/runtime/issues/74052#issuecomment-1218289577).

madelson commented on src/libraries/System.Private.CoreLib/src/System/Resources/ResourceReader.cs:
```diff
@@ -169,7 +170,7 @@ private unsafe void Dispose(bool disposing)
         }
 
-        internal static unsafe int ReadUnalignedI4(int* p)
+        private static unsafe int ReadUnalignedI4(int* p)
```
`ResourceReader` had a few methods marked as `internal` but not actually used outside of the class. I made them `private` to hopefully avoid these contributing to new thread-safety bugs in the future.

joperezr replied:
I assumed we also checked if this was used in System.Resources.Extesnions and their tests? That assembly also compiles this file, so it will be good to make sure.

madelson replied:
Confirmed that that project still builds and its tests pass. It includes RuntimeResourceSet but doesn't seem to reference it at all...

madelson commented on src/libraries/System.Private.CoreLib/src/System/Resources/RuntimeResourceSet.cs:
```diff
@@ -353,14 +355,11 @@ private IDictionaryEnumerator GetEnumeratorHelper()
 
-            // Normally calling LoadString or LoadObject requires
-            // taking a lock.  Note that in this case, we took a
-            // lock before calling this method.
```
This comment was not (no longer?) correct. We do typically call these within a lock, but it is a lock on either `_resCache` or `_caseInsensitiveTable`. That lock was previously making things thread-safe so long as the caller was only using `GetString/GetObject` and was not flipping `ResourceManager.IgnoreCase` on and off. 

Now, `LoadString`/`LoadObject` _do_ lock internally so there is no edge case here.

madelson commented on src/libraries/System.Resources.ResourceManager/tests/ResourceManagerTests.cs:
```diff
@@ -222,6 +225,46 @@ public static void IgnoreCase(string key, string expectedValue)
+            using Barrier barrier = new(Threads);
+            Task task = Task.WhenAll(Enumerable.Range(0, Threads).Select(_ => Task.Run(WaitForBarrierThenEnumerateResources)));
+
+            Assert.True(task.Wait(TimeSpan.FromSeconds(10)));
```
Before my fixes, this test reliably would reproduce both issues on my machine. Let me know if there is a better way to test for concurrency issues like this. 

I do think it is very important to have such a test, because this problem was actually previously [fixed in 2015](https://github.com/dotnet/coreclr/pull/75/files) and then broken again in 2020 (I think the break happened [here](https://github.com/dotnet/runtime/pull/44454/files#diff-21262dbd9abd6640e1b60d773746e46f6bb796efc8c4ddab9aa9be6f975b284bL318).

madelson commented on src/libraries/System.Resources.ResourceManager/tests/System.Resources.ResourceManager.Tests.csproj:
```diff
@@ -40,23 +48,14 @@
-    <Content Include="Resources\icon.ico"
-          Link="%(Filename)%(Extension)"
-          CopyToOutputDirectory="PreserveNewest"
-          Visible="false" />
```
Reformatting was done by VS

madelson commented on src/libraries/System.Resources.ResourceManager/tests/System.Resources.ResourceManager.Tests.csproj:
```diff
@@ -19,10 +24,13 @@
+    <EmbeddedResource Include="Resources\AToZResx.resx">
+      <Generator>ResXFileCodeGenerator</Generator>
+      <LastGenOutput>AToZResx.Designer.cs</LastGenOutput>
+    </EmbeddedResource>
```
This is what VS generated when I made a new resx file. Any reason to change it to match what the other files were using?

stephentoub replied:
It'd be nice to be consistent.

madelson committed:
Remove trailing whitespace

madelson commented on src/libraries/System.Private.CoreLib/src/System/Resources/ResourceReader.cs:
```diff
@@ -459,68 +464,74 @@ private unsafe string AllocateStringForNameIndex(int index, out int dataOffset)
-            }
-            else
+
+            lock (this)
```
> it seems like the _store is only ever initialized with a Stream that's under our control, right?

@stephentoub I don't think that's correct. There is a public `ResourceReader(Stream)` constructor that populates `_store` with the user `Stream`. Of course most users are probably going through `ResourceManager` / `RuntimeResourceSet` where the runtime owns the stream. 

However, when it comes to calling arbitrary user code while holding the lock, this class is _already doing that_ because locking on `this` is `ResourceReader`'s primary existing synchronization mechanism. For example, consider the following code:
```
void Main()
{
	var myStream = new MyStream();
	var reader = new ResourceReader(myStream);
	myStream.Reader = reader;
	// prints "Calling user code from within locked ResourceReader" many times
	foreach (DictionaryEntry entry in reader) { }
}

class MyStream : MemoryStream
{
	public ResourceReader Reader { get; set; }
	
	public MyStream() : base(GetBytes()) { }
	
	private static byte[] GetBytes()
	{
		using var stream = typeof(Expression).Assembly.GetManifestResourceStream("FxResources.System.Linq.Expressions.SR.resources");
		var bytes = new byte[stream.Length];
		if (stream.Read(bytes, 0, bytes.Length) != bytes.Length) { throw new Exception(); }
		return bytes;
	}
	
	public override int Read(byte[] buffer, int offset, int count)
	{
		if (this.Reader != null && Monitor.IsEntered(this.Reader)) 
		{ 
			Console.WriteLine("Calling user code from within locked ResourceReader");
		}
		return base.Read(buffer, offset, count);
	}
}
```

madelson replied:
Any thoughts with this @stephentoub ? My view is that since we were already using this locking approach merely extending it to be used consistently shouldn't be a concern. 

Today, `ResourceReader` is "mostly" thread-safe apart from a couple methods missing locks. If we wanted to avoid this, I think we'd need to remove all locking from `ResourceReader` and lock in `RuntimeResourceSet` instead to preserve thread-safety. This could be a breaking change for people relying on `ResourceReader`'s thread-safety, though.

joperezr commented on src/libraries/System.Private.CoreLib/src/System/Resources/ResourceReader.Core.cs:
```diff
@@ -55,6 +55,8 @@ internal ResourceReader(Stream stream, Dictionary<string, ResourceLocator> resCa
             "the user to only get one error.")]
         private object DeserializeObject(int typeIndex)
         {
+            Debug.Assert(Monitor.IsEntered(this));
```
Can we also add a comment explaining why this is being added?

madelson replied:
After double checking I don't think we need this one; this method doesn't access `_store` directly. Removed

joperezr commented on src/libraries/System.Private.CoreLib/src/System/Resources/ResourceReader.Core.cs:
```diff
@@ -94,6 +96,8 @@ private object DeserializeObject(int typeIndex)
             "Custom readers as well as custom objects on the resources file are not observable by the trimmer and so required assemblies, types and members may be removed.")]
         private bool InitializeBinaryFormatter()
         {
+            Debug.Assert(Monitor.IsEntered(this));
```
Same here. That way in the future people that look at this code will know why is the protection expected.

madelson replied:
Removed this one too. Not needed.

joperezr commented on src/libraries/System.Resources.ResourceManager/tests/ResourceManagerTests.cs:
```diff
@@ -222,6 +225,46 @@ public static void IgnoreCase(string key, string expectedValue)
+        [Theory]
+        [InlineData(false)]
+        [InlineData(true)]
+        public static void TestResourceManagerIsSafeForConcurrentAccessAndEnumeration(bool useEnumeratorEntry)
```
Does it make sense to add equivalent tests in System.Resources.Extensions as well and perhaps with a resource file that has something other than strings?

joperezr reviewed (commented):
Sorry for the delay in reviewing this @madelson, I've left some comments but overall this is looking good, thanks for the contribution.

madelson committed:
Address feedback from https://github.com/dotnet/runtime/pull/75054

madelson commented:
> I've left some comments but overall this is looking good

Thanks @joperezr . I've pushed a change addressing your feedback. Let me know if there is anything else to do.

joperezr commented:
>         private void SkipString()

What about `private void SkipString()`? That one seems to be using `_store` as well.

---
Refers to: src/libraries/System.Private.CoreLib/src/System/Resources/ResourceReader.cs:178 in 5d99bf6. [](commit_id = 5d99bf64f64719340f90f061c8d32aa07183ca23, deletion_comment = False)

joperezr commented:
Same for `private unsafe string AllocateStringForNameIndex(int index, out int dataOffset)` that one also seems to be accessing _store

madelson commented:
> What about private void SkipString()? That one seems to be using _store as well.

@joperezr this one can't have an assertion because it is called both during initialization (which per my comment on `_store` does not use locking) _and_ post-initialization.

**Do you think it is worth adding a comment like this to SkipString()?**:
```
// Note: this method assumes that it is called either from the constructor or within another method that locks on this
```

> Same for private unsafe string AllocateStringForNameIndex(int index, out int dataOffset) that one also seems to be accessing _store

This one accesses `_store` but note that the access is wrapped in a `lock(this)` block within that method. I only added assertions to methods that make assumptions about locking happening in the caller.

> There are some places where we still use _store and are not doing either of the above

I believe all of those places are called during initialization, which is called out by the comment on `_store`:

```
        // Backing store we're reading from. Usages outside of constructor
        // initialization must be protected by lock (this).
```

In cases which are _always_ called during initialization (e.g. `_ReadResources()`) I assert that the lock is _not_ taken.

joperezr commented:
I see thanks for the explanation and sorry that I missed some of those comments.

> Do you think it is worth adding a comment like this to SkipString()?:

Yes, I do think this would be good for consistency.

joperezr reviewed (approved):
Apart from my last NIT of just adding a comment to `SkipString` this change looks good to me. @stephentoub since you commented on this PR earlier, any thoughts on the reply from @madelson?

madelson committed:
Add comment in response to https://github.com/dotnet/runtime/pull/75054#issuecomment-1317470280

madelson commented:
Wanted to bump this in case it fell through the cracks.

stephentoub commented on src/libraries/System.Resources.Extensions/tests/BinaryResourceWriterUnitTest.cs:
```diff
@@ -500,6 +501,50 @@ public static void EmbeddedResourcesAreUpToDate()
             }
         }
 
+        [ConditionalTheory(typeof(PlatformDetection), nameof(PlatformDetection.IsBinaryFormatterSupported))]
+        /// <summary>
+        /// This test has multiple threads simultaneously loop over the keys of a moderately-sized resx using
+        /// <see cref="ResourceManager"/> and call <see cref="ResourceManager.GetString(string)"/> for each key.
+        /// This has historically been prone to thread-safety bugs because of the shared cache state and internal
+        /// method calls between RuntimeResourceSet and <see cref="DeserializingResourceReader"/>.
+        ///
+        /// Running with <paramref name="useEnumeratorEntry"/> TRUE replicates https://github.com/dotnet/runtime/issues/74868,
+        /// while running with FALSE replicates the error from https://github.com/dotnet/runtime/issues/74052.
+        /// </summary>
+        /// <param name="useEnumeratorEntry">
+        /// Whether to use <see cref="IDictionaryEnumerator.Entry"/> vs. <see cref="IDictionaryEnumerator.Key"/> when enumerating;
+        /// these follow fairly different code paths.
+        /// </param>]
+        [InlineData(false)]
+        [InlineData(true)]
```
```suggestion
        /// <summary>
        /// This test has multiple threads simultaneously loop over the keys of a moderately-sized resx using
        /// <see cref="ResourceManager"/> and call <see cref="ResourceManager.GetString(string)"/> for each key.
        /// This has historically been prone to thread-safety bugs because of the shared cache state and internal
        /// method calls between RuntimeResourceSet and <see cref="DeserializingResourceReader"/>.
        ///
        /// Running with <paramref name="useEnumeratorEntry"/> TRUE replicates https://github.com/dotnet/runtime/issues/74868,
        /// while running with FALSE replicates the error from https://github.com/dotnet/runtime/issues/74052.
        /// </summary>
        /// <param name="useEnumeratorEntry">
        /// Whether to use <see cref="IDictionaryEnumerator.Entry"/> vs. <see cref="IDictionaryEnumerator.Key"/> when enumerating;
        /// these follow fairly different code paths.
        /// </param>]
        [ConditionalTheory(typeof(PlatformDetection), nameof(PlatformDetection.IsBinaryFormatterSupported))]
        [InlineData(false)]
        [InlineData(true)]
```

madelson replied:
Moved the attribute.

stephentoub commented on src/libraries/System.Resources.Extensions/tests/BinaryResourceWriterUnitTest.cs:
```diff
@@ -500,6 +501,50 @@ public static void EmbeddedResourcesAreUpToDate()
+
+            const int Threads = 10;
+            using Barrier barrier = new(Threads);
+            Task task = Task.WhenAll(Enumerable.Range(0, Threads).Select(_ => Task.Run(WaitForBarrierThenEnumerateResources)));
```
We should probably avoid depending on the thread pool here, or else with the concurrency in unit tests these work items could be delayed arbitrarily long.
```suggestion
            Task task = Task.WhenAll(Enumerable.Range(0, Threads).Select(_ => Task.Factory.StartNew(WaitForBarrierThenEnumerateResources, CancellationToken.None, TaskCreationOptions.LongRunning, TaskScheduler.Default)));
```

madelson replied:
Changed to LongRunning task

stephentoub commented on src/libraries/System.Resources.Extensions/tests/BinaryResourceWriterUnitTest.cs:
```diff
@@ -500,6 +501,50 @@ public static void EmbeddedResourcesAreUpToDate()
+            using Barrier barrier = new(Threads);
+            Task task = Task.WhenAll(Enumerable.Range(0, Threads).Select(_ => Task.Run(WaitForBarrierThenEnumerateResources)));
+
+            Assert.True(task.Wait(TimeSpan.FromSeconds(10)));
```
Is 10 seconds going to be sufficient to avoid spurious timeouts in CI?

madelson replied:
Bumped it to 30s to be extra safe.

stephentoub commented on src/libraries/System.Resources.Extensions/tests/BinaryResourceWriterUnitTest.cs:
```diff
@@ -500,6 +501,50 @@ public static void EmbeddedResourcesAreUpToDate()
+            using Barrier barrier = new(Threads);
+            Task task = Task.WhenAll(Enumerable.Range(0, Threads).Select(_ => Task.Run(WaitForBarrierThenEnumerateResources)));
+
+            Assert.True(task.Wait(TimeSpan.FromSeconds(10)));
```
Why Task.WhenAll(...).Wait() rather than Task.WaitAll(...)?

stephentoub reviewed (approved):
A few comments, but otherwise LGTM. Thanks.

madelson committed:
Merge remote-tracking branch 'origin/main' into fix-74052-and-74868

madelson commented on src/libraries/System.Private.CoreLib/src/System/Resources/ResourceReader.cs:
```diff
@@ -1077,31 +1103,30 @@ public DictionaryEntry Entry
-                    { // locks should be taken in the same order as in RuntimeResourceSet.GetObject to avoid deadlock
-                        lock (_reader._resCache)
+                    // Lock the cache first, then the reader (in this case, we don't actually need to lock the reader and cache at the same time).
+                    // Lock order MUST match RuntimeResourceSet.GetObject to avoid deadlock.
```
Nope, we can add this check.

madelson commented on src/libraries/System.Resources.Extensions/tests/BinaryResourceWriterUnitTest.cs:
```diff
@@ -500,6 +501,50 @@ public static void EmbeddedResourcesAreUpToDate()
+                {
+                    object key = useEnumeratorEntry ? enumerator.Entry.Key : enumerator.Key;
+                    manager.GetObject((string)key);
+                    Thread.Sleep(1);
```
At least on my machine it does. Just ran again 3 times and got a failure each time without the fix.

madelson committed:
Implement feedback from https://github.com/dotnet/runtime/pull/75054

madelson commented on src/libraries/System.Resources.ResourceManager/tests/ResourceManagerTests.cs:
```diff
@@ -222,6 +225,46 @@ public static void IgnoreCase(string key, string expectedValue)
+
+            const int Threads = 10;
+            using Barrier barrier = new(Threads);
+            Task task = Task.WhenAll(Enumerable.Range(0, Threads).Select(_ => Task.Run(WaitForBarrierThenEnumerateResources)));
```
Doen

madelson commented:
@joperezr looking at the new build failures these seem unrelated? https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-75054-merge-b07193f9df6f4ac4a6/JIT.1/1/console.735a06a0.log?helixlogtype=result what do you think?

jkotas commented:
The new failure is https://github.com/dotnet/runtime/issues/79517

madelson committed:
Merge branch 'dotnet:main' into fix-74052-and-74868

madelson commented:
@joperezr anything else to do here?

buyaa-n commented:
/backport to release/6.0

joperezr commented:
@buyaa-n assuming you also plan to backport to 7.0?

buyaa-n commented:
/backport to release/7.0
